src/Cache/ClientCache.php

```diff
@@ -0,0 +1,48 @@
+<?php
+declare(strict_types = 1);
+
+namespace LanguageServer\Cache;
+
+use LanguageServer\LanguageClient;
+use Sabre\Event\Promise;
+
+/**
+ * Caches content through a xcache/* requests
+ */
+class ClientCache implements Cache
+{
+    /**
+     * @param LanguageClient $client
+     */
+    public function __construct(LanguageClient $client)
+    {
+        $this->client = $client;
+    }
+
+    /**
+     * Gets a value from the cache
+     *
+     * @param string $key
+     * @return Promise <mixed>
+     */
+    public function get(string $key): Promise
+    {
+        return $this->client->xcache->get($key)->then('unserialize')->otherwise(function () {
+            // Ignore
+        });
+    }
+
+    /**
+     * Sets a value in the cache
+     *
+     * @param string $key
+     * @param mixed  $value
+     * @return Promise
+     */
+    public function set(string $key, $value): Promise
+    {
+        return $this->client->xcache->set($key, serialize($value))->otherwise(function () {
```

Comment:
you can either do stuff on your own:
https://3v4l.org/5Y60r

or use libs which provide this kind of functionality, e.g:
http://jmsyst.com/libs/serializer

Re Security: your class doesnt know whether the content beeing stored is provided by user-input or not.

> That's why all cache items are prefixed with a version. The cache format will eventually change at some point, whether I use JSON or serialize, it needs to be versioned.
nice!